README.md
# room_reservation_meteor

This is an individual assignment.  This 60 point mini-project is to earn the 60 points you missed out on by having already completed the 3 code school courses in CY355, as well as provide a "more gentle" introduction to some of the concepts you'll be observing in the Roborally meteor game used for Milestones 6 and 7.  You are to extend the functionality of the new CS403 Meteor Room Booking Program (MRBP) provided in the <GITHUB REPO LINK HERE>.  Clone this repository and get the meteor project up and running.

This program is intended to allow different users to collaborate and share a common room.  The current functionality of the room reservation system is as follows:

- This program assumes the current year is 2018 and does not support adding dates for other years.
- In order to see the room reservation system, you must be logged into the system.
- Once logged into the system, you see the next five reservations for room 101 (assuming there are any).
- You can create a new reservation by filling out the event description and using the drop down boxes to select the month, day, start time, and end time of your event.
- The room is booked for the user currently signed in.
- The form for adding new reservations has some logic built into it.  Firstly, they user must have all fields filled out for the reservation to complete.  Also, the user is not allowed to have an end date for the reservation which is before their chosen start date.

Extend the code by adding the following additional functionality:

- Currently, room 101 is hard coded into program, and this program only supports tracking events for one room.  Extend the program to allow users to make reservations for 8 different rooms.  After signing in, the user will be allowed to choose which room they would like to reserve (use rooms 101-108).  After selecting the room they would like to reserve, they will be allowed to use the provided reservation form to add a new event for this room.
- The "next 5 upcoming event" section currently displays all event reservations.  Since you now have 8 rooms you're reserving for, make sure your upcoming events section only displays events for the selected room.
- There is currently no protection against double booking the room.  The user should not be allowed to reserve the room during a time which overlaps with an existing reservation.  Make sure you show a warning message to the user, similar to what has already been implemented on the form if they haven't filled it out completely.
- The start and end time fields in the "next 5 upcoming events" section is displayed using JavaScript's built in Date objects.  Format these fields in a more easily readable format.  Feel free to use something like moment.js to assist you with this task.
- Customize the look of your application by changing .css properties.  Make at least five modifications to the .css so that you like the look of your room reservation page.  Be creative!

A signed copy of the coversheet is due beginning class on 23/24 April (WEB-6) along with any appropriate documentation of assistance.  You will submit your final code by the beginning of class to your turn-in folder (\\usmasvddeecs\eecs\Cadet\Courses\CS403\18-2\<YOUR FOLDER>).  To receive a grade on your mini-project, you will schedule a 15 minute window of time with your instructor to go over your project improvements individually.  Make sure you get your meteor project running on your laptop of on a lab machine prior to your scheduled window of time.

Again, this is an individual assignment, so make sure you cite any help you have received.


History.txt
17 April
  added buttons for each room

19 April
  must add in drop down for room number (todo)
    changed html, add roomNumberDropdown
    add rooms function in startEndSelectors.helpers
    add check to see if room dropdown filled out before submitting
    submit with respective room number, not 101 default
  must show only one room's reservations (todo)
    done in main.js
      upcomingReservations.helpers with mongo query
  (canceled) reroute room number buttons to own respective pages
  find way to change reservation page heading to match which button was pressed
    done with global JS function and variable

    USE AN OUTSIDE FUNCTION AND RUN INSIDE THE CORRECT TEMPLATE HELPER


  also link button pressed id with which reservations are shown, currently hardcoded 101


imports/api/reservations.js
import { Mongo } from 'meteor/mongo';
import { check } from 'meteor/check';

export const Reservations = new Mongo.Collection('reservations');

Meteor.methods({
    // Provides a meteor method which inserts and event into the database.
    // Validation is expected to have already occurred on this event by the time this method is called.
    "insertEvent" : function(eventDescriptionText, reservationStartTime, reservationEndTime, roomNumber) {
        check(eventDescriptionText, String);
        check(reservationStartTime, Date);
        check(reservationEndTime, Date);
        
        Reservations.insert({
          eventDescription: eventDescriptionText,
          createdAt: new Date(),
          reservationStart: reservationStartTime,
          reservationEnd: reservationEndTime,
          owner: Meteor.userId(),
          username: Meteor.user().username,
          room: roomNumber
        });
    }
});

client/main.js
import { Template } from 'meteor/templating';
import { ReactiveVar } from 'meteor/reactive-var';
import { Meteor } from 'meteor/meteor';

import './main.html';
import '../imports/startup/accounts-config.js';
import {Reservations} from "../imports/api/reservations.js";

// Some helper functions created to make life easier later on.
// Checks if an object is empty.  Returns true if the object is empty, false otherwise.
function isEmpty(obj) {
  for(var key in obj) {
    if(obj.hasOwnProperty(key))
        return false;
  }
  return true;
}

// Left-Zero pads a number to a given width.  Returns a string
function zeroFill( number, width ) {
  width -= number.toString().length;
  if ( width > 0 ) {
    return new Array( width + (/\./.test( number ) ? 2 : 1) ).join( '0' ) + number;
  }
  return number + ""; // always return a string
}

// Returns an array containing the numbers of days in a given month.
function getDaysForMonth(month) {
  var days = [];

  if (month == "January" || month == "March" || month == "May" || month == "July" || month == "August" || month == "October" || month == "December") {
    days = [];
    for (i=1; i<=31; i++) {
      days.push(i);
    }
  }
  else if (month == "February") {
    days = [];
    for (i=1; i<=28; i++) {
      days.push(i);
    }
  }
  else if (month == "April" || month == "June" || month == "September" || month == "November") {
    days = [];
    for (i=1; i<=30; i++) {
      days.push(i);
    }
  }
  else {
    days = [];
  }

  return days;
}

var currRoomNum = "101"

// Used for populating the both month selection fields.
Template.startEndSelectors.helpers({
  months: function(){
    return ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
  },
  times: function(){
    var times = []
    for (hour=0; hour<=23; hour++) {
      for (minute=0; minute<60; minute+=15) {
        times.push(zeroFill(hour, 2) + zeroFill(minute, 2));
      }
    }
    return times
  },
  rooms: function(){
    var rooms = []
    for (room=101; room<=108; room++) {
      rooms.push(room);
    }
    return rooms
  }
});

// Detects when the month has changed and populates the correct number of days for that month into the dayDropdown select box.
Template.startEndSelectors.events({
  // When they select a month, populate the correct number of days in in the dayDropdown select.
  "change #startMonthDropdown": function (event, template) {
    var days = getDaysForMonth($(event.currentTarget).val())

    // clear the dayDropdown select options
    $('#startDayDropdown')
      .find('option')
      .remove()
      .end()
      .append('<option>Select Day</option>');

    // repopulate with correct number of days
    if (!isEmpty(days)) {
      $.each(days, function(key, value) {
        $('#startDayDropdown').append($("<option/>", {
          value: key+1,
          text: value
        }));
      });
    }
  },
  "change #endMonthDropdown": function (event, template) {
    var days = getDaysForMonth($(event.currentTarget).val())

    // clear the dayDropdown select options
    $('#endDayDropdown')
      .find('option')
      .remove()
      .end()
      .append('<option>Select Day</option>');

    // repopulate with correct number of days
    if (!isEmpty(days)) {
      $.each(days, function(key, value) {
        $('#endDayDropdown').append($("<option/>", {
          value: key+1,
          text: value
        }));
      });
    }
  }
});

// Sets the feedback message initially to be "all fields must be completed."
Template.makeReservationBox.onCreated(function() {
	this.eventFeedbackMsg = new ReactiveVar("All fields must be complete");
});

// updates the event feedback message if need be.
Template.makeReservationBox.helpers({
  eventFeedbackMsg: function() {
	return Template.instance().eventFeedbackMsg.get();  },
  checkIfInUse: function() {
	return false;
  }
});

// Logic for the form.  Ensures all fields have been filled out.  Ensures the start date is before the end date.
Template.makeReservationBox.events({
  "click #submitReservationButton" : function(event, template) {
    var eventDescription = $("#eventDescription").val()

    var startMonth = $('#startMonthDropdown').val();
    var startDay = parseInt($('#startDayDropdown').val());
    var startTime = $('#startTimeDropdown').val();

    var endMonth = $('#endMonthDropdown').val();
    var endDay = parseInt($('#endDayDropdown').val());
    var endTime = $('#endTimeDropdown').val();

    // Validate they have filled out all the boxes
    if (eventDescription == "") {
  		template.eventFeedbackMsg.set("You must enter an: Event Description");
  		template.find('.eventFeedback').style.color="#ff4444";
      return
    }

    else if(startMonth==null || startDay==null || startTime == null || isNaN(startDay) ) {
  		template.eventFeedbackMsg.set("You must enter a complete Event Start date");
  		template.find('.eventFeedback').style.color="#ff4444";
      return
    }
    else if(endMonth==null || endDay==null || endTime == null || isNaN(endDay)) {
  		template.eventFeedbackMsg.set("You must enter a complete Event End date");
  		template.find('.eventFeedback').style.color="#ff4444";
      return
    }

    //They have, so validate startTime is before endTime
    else {
      var startTimeString = startMonth.toString() + " " + startDay.toString() + ", 2018 " + startTime.toString().substring(0,2) + ":" + startTime.toString().substring(2,4) + ":00";
      var reservationStart = new Date(startTimeString);

      var endTimeString = endMonth.toString() + " " + endDay.toString() + ", 2018 " + endTime.toString().substring(0,2) + ":" + endTime.toString().substring(2,4) + ":00";
      var reservationEnd = new Date(endTimeString);

      if (reservationStart >= reservationEnd) {
	      template.eventFeedbackMsg.set("End date must be after start date");
	      template.find('.eventFeedback').style.color="#ff4444";
      }
      /*
      else if (template.checkIfInUse){
	      template.eventFeedbackMsg.set("Room is in use at this time");
	      template.find('.eventFeedback').style.color="#ff4444";
      }
      */
      else {
        var overlap = false;
// https://stackoverflow.com/questions/12956438/accessing-mongodb-collection-values-in-javascript
// 2018 APR 22. Assistance given to the author, stackoverflow post
// I used the information in the third post down that uses ForEach to iterate through a collection.
// I was using a for loop to manually iterate through.
// But it wasn't working because it probably only returned the cursor, not the actual object
// This allowed me to actually compare the date objects and check for overlapping times. West Point, NY
        Reservations.find({"room": currRoomNum}, { sort: { reservationStart: 1 }}).forEach(function(obj) {
          var existStart = obj.reservationStart;
          var existEnd = obj.reservationEnd;
          if ((existStart <= reservationStart && existEnd >= reservationStart) ||
            (existStart <= reservationEnd && existEnd >= reservationEnd)) {
              overlap = true;
            }
        })
        if (overlap) {
          template.eventFeedbackMsg.set("Room is in use at this time");
  		    template.find('.eventFeedback').style.color="#ff4444";
          console.log("Event overlaps");
        }
        else {
          Meteor.call("insertEvent", eventDescription, reservationStart, reservationEnd, currRoomNum);
  		    template.eventFeedbackMsg.set("Event entered successfully");
  		    template.find('.eventFeedback').style.color="#44ff22";
          console.log("Event entered");
        }
        /////////////////////////////////////////////////////////////////////////////////////////////
      }
    }
  }
});


// Locates and returns the next five events which are occurring.
Template.upcomingReservations.helpers({
  reservations : function() {
    var res = Reservations.find({"room": currRoomNum}, { sort: { reservationStart: 1 }}).fetch();

    if (res.length > 5) {
      return res.slice(0,5);
    }
    else {
      return res;
    }
  },
  // apparently this is super necessary here because these must be functions, not variables
  roomNum : function() {
    return currRoomNum
  }
});

Template.reservationPage.events({
  "click #backButton" : function() {
    console.log("back button clicked");
    Router.go("/")
  }
});

Template.roomSelectPage.events({
  'click #room101': function() {
    console.log("101 clicked!");
    currRoomNum = "101";
    Router.go("/reservationPage");
  },
  'click #room102': function() {
    console.log("102 clicked!");
    currRoomNum = "102";
    Router.go("/reservationPage")
  },
  'click #room103': function() {
    console.log("103 clicked!");
    currRoomNum = "103";
    Router.go("/reservationPage")
  },
  'click #room104': function() {
    console.log("104 clicked!");
    currRoomNum = "104";
    Router.go("/reservationPage")
  },
  'click #room105': function() {
    console.log("105 clicked!");
    currRoomNum = "105";
    Router.go("/reservationPage")
  },
  'click #room106': function() {
    console.log("106 clicked!");
    currRoomNum = "106";
    Router.go("/reservationPage")
  },
  'click #room107': function() {
    console.log("107 clicked!");
    currRoomNum = "107";
    Router.go("/reservationPage")
  },
  'click #room108': function() {
    console.log("108 clicked!");
    currRoomNum = "108";
    Router.go("/reservationPage")
  }
});

// https://stackoverflow.com/questions/22087907/how-to-format-date-in-meteor-template
// 2018 APR 22 Assistance given to author. stackoverflow post
// I used the top answer to use a function to format my string using moment.
// I also learned how to use functions in the spacebars format.
// This allowed me to display readable dates/times on my list of reservations.
Template.eventLine.helpers({
  dateFormat : function(date) {
    return moment(date).format('LLL');
  }
});
